be/src/runtime/jdbc_driver_manager.cpp

```diff
@@ -0,0 +1,264 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#include "runtime/jdbc_driver_manager.h"
+
+#include <atomic>
+#include <boost/algorithm/string/predicate.hpp> // boost::algorithm::ends_with
+#include <chrono>
+
+#include "env/env.h"
+#include "fmt/format.h"
+#include "gutil/strings/split.h"
+#include "http/http_client.h"
+#include "util/defer_op.h"
+#include "util/dynamic_util.h"
+#include "util/file_utils.h"
+#include "util/md5.h"
+#include "util/slice.h"
+
+namespace starrocks {
+
+struct JDBCDriverEntry {
+    JDBCDriverEntry(const std::string& name_, const std::string& checksum_) : name(name_), checksum(checksum_) {}
+
+    ~JDBCDriverEntry();
+
+    inline bool is_expected(const std::string& name, const std::string& checksum) {
+        return this->name == name && this->checksum == checksum;
+    }
+
+    std::string name;
+    std::string checksum;
+    int64_t first_access_ts;
+    // ${driver_dir}/${name}_${checksum}_${first_access_ts}.jar
+    std::string location;
+
+    std::atomic<bool> is_available{false};
+
+    std::atomic<bool> should_delete{false};
+
+    std::mutex download_lock;
+    bool is_downloaded = false;
+};
+
+JDBCDriverEntry::~JDBCDriverEntry() {
+    if (should_delete.load()) {
+        LOG(INFO) << fmt::format("try to delete jdbc driver {}", location);
+        FileUtils::remove(location);
+    }
+}
+
+JDBCDriverManager::JDBCDriverManager() {}
+
+JDBCDriverManager::~JDBCDriverManager() {
+    std::unique_lock<std::mutex> l(_lock);
+    _entry_map.clear();
+}
+
+JDBCDriverManager* JDBCDriverManager::getInstance() {
+    static std::unique_ptr<JDBCDriverManager> manager;
+    if (manager == nullptr) {
+        manager.reset(new JDBCDriverManager());
+    }
+    return manager.get();
+}
+
+Status JDBCDriverManager::init(const std::string& driver_dir) {
+    std::unique_lock<std::mutex> l(_lock);
+
+    _driver_dir = driver_dir;
+    RETURN_IF_ERROR(FileUtils::create_dir(_driver_dir));
+    std::vector<std::string> driver_files;
+    RETURN_IF_ERROR(FileUtils::list_files(Env::Default(), _driver_dir, &driver_files));
+    // load jdbc drivers from file
+    for (auto& file : driver_files) {
+        std::string target_file = fmt::format("{}/{}", _driver_dir, file);
+        // remove all tmporary files
+        if (boost::algorithm::ends_with(file, TMP_FILE_SUFFIX)) {
```

Comment:
std lib no this method？

Reply:
> std lib no this method？

`std::basic_string<>::ends_with` was introduced in c++20, but we are using c++17